### PR TITLE
Clever Git

### PR DESCRIPTION
**Pull request by DuncanFenning**
### Blog title
Clever Git

### Blog description
How to improve your git usage to make branch commits/history cleaner

#### Author
- [X] I have checked for grammatical errors
- [X] I have checked for spelling errors
- [X] All information (technical or otherwise) is correct to my knowledge

#### Editor 1
- [x] I have checked for grammatical errors
- [x] I have checked for spelling errors
- [x] All information (technical or otherwise) is correct to my knowledge

#### Editor 2
- [x] I have checked for grammatical errors
- [x] I have checked for spelling errors
- [x] All information (technical or otherwise) is correct to my knowledge

### CODE REVIEW

File: _posts/2015-10-22-clever-git.md [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+The obvious way that most people use to get around this problem is with a `git merge`. This updates your featureBranch with the latest master branch. However, every time you merge master, you will add a new commit to your branch history. Doing this once may not seem like it could be a problem but when you are working on a branch for a prolonged period of time, this can occur multiple times, leaving your commit history in a mess! This in turn makes it difficult for other developers to understand the history of your branch.
+
+Instead of doing a `git merge master into your branch, you can instead do `git rebase master`. This has the same affect as 'merging master' but instead of adding a new commit to your featureBranch, it re-writes the project history by re-writing your commits. This leaves you with a cleaner project history.
```
Comment (hxpaul) [inline _posts/2015-10-22-clever-git.md]: missing a backtick after `git merge master`

**Comment by hxpaul** [comment]
A nice article :+1:

File: _posts/2015-10-22-clever-git.md [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+Instead of doing a `git merge master` into your branch, you can instead do `git rebase master`. This has the same affect as 'merging master' but instead of adding a new commit to your featureBranch, it re-writes the project history by re-writing your commits. This leaves you with a cleaner project history.
+
+To improve your commit history even more, you can clean the commits on your featureBranch by doing `git rebase -i master`. This gives you many options for every commit you have made on your featureBranch. It lets you squash commits together and reword your commit messages. This can be used to group specific changes into individual commits and squash small commits into other commits, keeping your branch history concise.
```
Comment (pmcnr-hx) [inline _posts/2015-10-22-clever-git.md]: "by doing `git rebase -i master`" :point_right: "by using the iterative rebase command `git rebase -i master`", no?

File: _posts/2015-10-22-clever-git.md [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+The obvious way that most people use to get around this problem is with a `git merge`. This updates your featureBranch with the latest master branch. However, every time you merge master, you will add a new commit to your branch history. Doing this once may not seem like it could be a problem but when you are working on a branch for a prolonged period of time, this can occur multiple times, leaving your commit history in a mess! This in turn makes it difficult for other developers to understand the history of your branch.
+
+Instead of doing a `git merge master` into your branch, you can instead do `git rebase master`. This has the same affect as 'merging master' but instead of adding a new commit to your featureBranch, it re-writes the project history by re-writing your commits. This leaves you with a cleaner project history.
```
Comment (pmcnr-hx) [inline _posts/2015-10-22-clever-git.md]: Make clear here the caveat that this is a great thing to do while your branch only exists locally in your working copy but you shouldn't do it if you have already pushed your branch publicly.

Comment (pmcnr-hx) [inline _posts/2015-10-22-clever-git.md]: A link [here](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) somewhere in the paragraph might also be useful.

File: _posts/2015-10-22-clever-git.md [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+# Interactive staging
+
+If asked to stage a file, most people will do: `git add path/to/file`. This stages all the changes made to that one file.
+
+Other people might do: `git add -p path/to/file`. Even better, you can see hunks of changes which you want to stage for each file, rather than blindly staging a whole file's changes.
+
+However not as many people know of: `git add -i`. This opens up more options for your changes. It lets you hand pick what changes you want to stage, see the diff for each file or revert certain file changes, all in once centralised place.
+
+Using interactive staging of files, along with rebasing master will leave you and your projects in a happy place.
```
Comment (pmcnr-hx) [inline _posts/2015-10-22-clever-git.md]: A link [here](https://git-scm.com/book/en/v2/Git-Tools-Interactive-Staging) somewhere in the section might be useful.

File: _posts/2015-10-22-clever-git.md [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+[Atlassian](https://www.atlassian.com/git/tutorials/merging-vs-rebasing/workflow-walkthrough) have an in depth tutorial about when to merge and when to rebase.
+
+# Interactive staging
+
+If asked to stage a file, most people will do: `git add path/to/file`. This stages all the changes made to that one file.
+
+Other people might do: `git add -p path/to/file`. Even better, you can see hunks of changes which you want to stage for each file, rather than blindly staging a whole file's changes.
+
+However not as many people know of: `git add -i`. This opens up more options for your changes. It lets you hand pick what changes you want to stage, see the diff for each file or revert certain file changes, all in once centralised place.
```
Comment (pmcnr-hx) [inline _posts/2015-10-22-clever-git.md]: "know of" :point_right: "know of interactive staging"?

**Comment by pmcnr-hx** [comment]
Excellent effort: :hand: (that's an high5)!

**Comment by DuncanFenning** [comment]
All updated @pmcnr-hx :D

**Comment by jackdcrawford** [comment]
Reviewing :book:

File: _posts/2015-10-22-clever-git.md [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+git checkout featureBranch
+-- some work goes on here --
+git add changes
+git commit changes
+git push changes
+```
+
+This would then go to peer review before being merged to master. However, when multiple people are all contributing to the same project, the workflow just isn't that simple. With multiple developers branching off master and multiple merges to master being done every day, branches can easily have forked history. This means a branch has been merged with master since you created your featureBranch, leaving your featureBranch with an out of date master branch.
+
+The obvious way that most people use to get around this problem is with a `git merge`. This updates your featureBranch with the latest master branch. However, every time you merge master, you will add a new commit to your branch history. Doing this once may not seem like it could be a problem but when you are working on a branch for a prolonged period of time, this can occur multiple times, leaving your commit history in a mess! This in turn makes it difficult for other developers to understand the history of your branch.
```
Comment (jackdcrawford) [inline _posts/2015-10-22-clever-git.md]: It would be clearer to add `git merge master` here, rather than just `git merge`

**Comment by jackdcrawford** [comment]
Really nice read @DuncanFenning, nice work :+1:

**Pull request merged**